### PR TITLE
Gnosis safe staging owner actions

### PR DESCRIPTION
**Pull request by jacko125**
Gnosis safe staging owner actions

TODO

- [x]  stage owner-actions.json as well to the protocol proxy. 

- [X] Check Gnosis safe API hasn't already got matching `to` and `data` and `!isExecuted`  of the preceding operation (otherwise confirm to skip).

### CODE REVIEW

File: publish/src/Gnosis-safe.abi.json [hunk 1]
```diff
@@ -0,0 +1,585 @@
+[
```
Comment (justinjmoses) [inline publish/src/Gnosis-safe.abi.json]: I really don't think we should be including this ABI in our repo. I think downloading this from Etherscan at deploy time is a reasonable approach. 

> As an aside, I have been thinking more and more of a better caching layer for ABI fetching. Seems silly to hit Etherscan's API for an immutable fetch. I've been thinking about creating an ABI service layer that uses Netlify for requests, GitHub as ABI storage, and Etherscan for the source of truth. Should be pretty simple to setup. Thoughts?

File: publish/src/commands/owner.js [hunk 2]
```diff
@@ -48,22 +81,38 @@ const owner = async ({ network, newOwner, deploymentPath }) => {
...
+		gray(
+			'Skipping - Running through operations during deployment that couldnt complete as not owner.'
+		)
 	);
 
-	for (const [key, entry] of Object.entries(ownerActions)) {
-		const { action, link, complete } = entry;
-		if (complete) continue;
+	// new owner should be gnosis safe proxy address
+	const protocolDaoContract = getSafeInstance(web3, newOwner);
```
Comment (justinjmoses) [inline publish/src/commands/owner.js]: I think you should consider moving the safe logic to the safe file entirely and keep this owner script clean

File: publish/src/commands/owner.js [hunk 3]
```diff
@@ -48,22 +81,38 @@ const owner = async ({ network, newOwner, deploymentPath }) => {
...
+		yellow(`Using Protocol DAO Safe contract at ${protocolDaoContract.options.address} `)
+	);
 
-		await confirmOrEnd(
-			yellow('YOUR TASK: ') + `Invoke ${bgYellow(black(action))} (${key}) via ${cyan(link)}`
-		);
+	console.log(gray('Looking for contracts whose ownership we should accept'));
 
-		entry.complete = true;
-		fs.writeFileSync(ownerActionsFile, stringify(ownerActions));
```
Comment (justinjmoses) [inline publish/src/commands/owner.js]: We should also remove the owner script entirely then - or else compile it during the deploy and remove it once all transactions are staged at the end of the deploy (to handle restarting deploys)

File: publish/src/safe-utils.js [hunk 4]
```diff
@@ -12,6 +11,52 @@ const DELEGATE_CALL = 1;
...
+		outputs: [],
+		payable: false,
+		stateMutability: 'nonpayable',
+		type: 'function',
+	},
+	{
+		constant: true,
+		inputs: [],
+		name: 'nonce',
+		outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
```
Comment (justinjmoses) [inline publish/src/safe-utils.js]: all we need is `name` and `inputs` effectively

File: publish/src/commands/owner.js [hunk 5]
```diff
@@ -81,46 +93,126 @@ const owner = async ({
...
+		});
 
-	// new owner should be gnosis safe proxy address
-	const protocolDaoContract = getSafeInstance(web3, newOwner);
-	// get protocolDAO nonce
-	const currentSafeNonce = await getSafeNonce(protocolDaoContract);
+		if (existingTx) {
+			console.log(
+				gray(
+					`Existing pending tx already submitted to gnosis safe - target address: ${target} and data: ${encodedData}`
```
Comment (justinjmoses) [inline publish/src/commands/owner.js]: This logic should be moved over to the safe script - this is the owner script, safe-specific code should be elsewhere.

Comment (jacko125) [reply publish/src/commands/owner.js]: Will have a look at refactoring more of this logic into safe-utils script.

**Pull request merged**